### DISCUSSION TITLE
Enabling Github Environments on the workflow level

### DISCUSSION THREAD

**Discussion by michhyun1** [community/community#14417] [Actions]
It seems as if I am only able to enable github environments on the Job level. I want to deploy my environment once and have my entire workflow run under that environment.

My use case is the following: 

I have a CI workflow that includes one workflow and ~10 jobs (and more will be added soon). Every single job requires a Github Secret. When a pull request is submitted to our main branch, GHA should require a manual deployment to our github environment called `CI Workflow` and once the approval is done - trigger the workflow and run the 10 jobs.

However, since the environments is only on the job level, I have to attach the `job.job-id.environment` property to each job. That means, when an external contributor submits a PR against our main branch, it looks like this:

![image](https://user-images.githubusercontent.com/54328451/162033517-839152ad-f0aa-485b-9690-e4312cd7b1cb.png)
[The PR is here](https://github.com/aws/amazon-chime-sdk-js/pull/2129)

Is there a way to deploy the environment only one time, and then run all of the jobs under that environment?

**Comment by chrispat** [comment]
@michhyun1 Could you provide some additional information on your scenario?  From what I see it feels like you are using environments with approvals and secrets to support a scenario that we have not thought about, but I am not 100% sure exactly what that is.

**Reply by michhyun1** [reply]
Sure - we want to use github environments as a way to unblock external contributor PRs to be able to run our Github Actions. The reason why we cannot do that today is because every Job in our Github Action workflow requires Github Secrets.

When an external contributor submits a PR against our repo, it will trigger our workflow, but the runner wont have access to our Github Secrets. 

As a workaround, we want to enable github environments on workflow level, so that an internal contributor can manually approve the deployment of the Github workflow in an environment. That way, the github workflow can run in an environment that has access to the secrets, and thereby pass the requirements for the PR to have these jobs pass before we can merge in the PR.

**Reply by michhyun1** [reply]
Any suggestion on the above? @chrispat

**Reply by rohit-gohri** [reply]
I have a very similar usecase/workflow. I have a monorepo that deploys multiple services to GCP across 3 environments (dev/stage/prod). 
The deployments for all the services happen in parallel across 5 different jobs in single workflow. Me and my team treat that as a single deployment, but Github treats it as n different deployments. I think it makes more sense to just group all those deployments as a single event in the UI and Deployments page.

**Reply by chrispat** [reply]
@michhyun1 Interesting scenario.  I don't think that will work without other changes as we won't send secrets to runs triggered from forks or external contributors that run against the code from that PR.  We do have the `pull_request_target` event which runs on the PR event but uses the code from the target branch rather than the PR branch.

@rohit-gohri That is good feedback.  There isn't anything you can do today to support that scenario but it is something we can consider for the future.

**Reply by michhyun1** [reply]
I thought a PR that is deployed to an Environment should have access to the Environment's secrets, regardless of whether or not it is a PR from a forked repo or not?

I plan to attach my secrets to an Environment, then deploy the external contributor PR to the Environment (once someone manually approves the deployment). The PR should have access to the Environment's secrets, correct?

**Reply by chrispat** [reply]
@michhyun1 at the moment we always treat PRs from fork the same regardless when using the `pull_request` event as it will checkout the code from the PR by default.  You could make it work with the `pull_request_target` event but you will need to be very careful in how you execute code from the untrusted actor.

**Reply by chrisponton** [reply]
+1 here too.

I am having this issue after implementing 10+ reusable workflows in a mono repo to be able to deploy to multiple environments.

I would like only the caller workflow to request approval, which means adding the environment property to this. But each called workflow needs access to different secrets based on the environment. If I pass the environment down and set it in the next jobs, the workflow picks up the correct secrets for that environment deploy, but I get numerous requests for approval, which means my pipeline doesn't flow like I want.

Any news on this?

**Reply by ccolella-mdc** [reply]
> +1 here too.
> 
> 
> 
> I am having this issue after implementing 10+ reusable workflows in a mono repo to be able to deploy to multiple environments.
> 
> 
> 
> I would like only the caller workflow to request approval, which means adding the environment property to this. But each called workflow needs access to different secrets based on the environment. If I pass the environment down and set it in the next jobs, the workflow picks up the correct secrets for that environment deploy, but I get numerous requests for approval, which means my pipeline doesn't flow like I want.
> 
> 
> 
> Any news on this?

My organization has this exact same issue. Either being able to set the environment on the called reusable workflow or on the job that calls the reusable workflow would solve it for us.

**Comment by HariSekhon** [comment]
I am also finding this frustrating... because I need to protect the deployment environments with reviews and when the workflow is consisting of different jobs that need access to the environment secrets, the workflow requires multiple approvals to get it to proceed, which is a waste of time.

I want deployment environment to be able to be configured at the workflow level so that once I have approved the environment deployment all the jobs in that workflow run should progress without repeatedly prompting me for every job in the flow.

Right now I've had to rewrite a relatively large 450 line workflow of multiple jobs which appeared nicely in the UI with different job steps into a single monolithic job to avoid this multiple approvals issue due to the lack of being able to set the deployment environment at the workflow level and approve it once in the UI.

On a related note, deployment environment support should be added to the reusable workflow level too, see:

https://github.com/github-community/community/discussions/18085

**Reply by dkirrane** [reply]
+1 I'm looking for same. 
And it would be good if reusable workflow could just `inherit` both the `environment` and `concurrency` from the main workflow.

e.g.
```
name: Stage

on:
  pull_request:
    branches:
      - main

concurrency: stage_environment

environment:
  name: stage_environment

jobs:
      
  call-workflow:
    uses: octo-org/example-repo/.github/workflows/workflow-cicd.yml@main
    secrets: inherit
    concurrency: inherit
    environment: inherit
```

**Reply by HariSekhon** [reply]
The concurrency I think is ok because it can lock it at the calling workflow level.

But the lack of environment pass-through is a clear gap in the lack of maturity of reusable workflows.

**Comment by HariSekhon** [comment]
This is such a tedious issue I've just had to rewrite a non-trivial 450 line workflow into a single job to minimize it to one approval for the workflow since each approval takes at least 3 clicks per job otherwise and wastes time waiting for approvals at each job/stage too.

You can't even automate around this with GitHub CLI, nor even with the Rest API, so I've just raised a request for them too:

https://github.com/cli/cli/issues/5773

https://github.com/github-community/community/discussions/18088

**Reply by Hronom** [reply]
Hello, same here especially useful for GitOps flow. When you wanna not start docker image build until user not approve this.
So I need provide possibility for the user to approve docker build and deploy and make sure that only one deployment is created.

**Comment by mkatanski** [comment]
Hello. Any news on that? Its the end of 2023 and I also have the same problem. I consider one workflow with multiple jobs as deployment and it would be nice to assign environment on workflow level.

**Comment by Hronom** [comment]
Hello GitHub, can you please find time and check this functionality in GitLab(it's your competitors). Hope you get inspiration. Thanks!

**Reply by gionn** [reply]
https://docs.gitlab.com/ee/ci/environments/#access-an-environment-for-preparation-or-verification-purposes

**Comment by e965** [comment]
It's very strange that something like this wasn't done from the beginning. At the moment it's the only thing stopping me from using Environments.

**Comment by ltrung** [comment]
Any update on this? This causes frustration for our external contributors.

**Comment by amrod** [comment]
I have the same issue. It's incredible this isn't at the top of Github's TODO list.

**Comment by DanielShepherd-RL** [comment]
Would love to see this functionality.

**Comment by peterramsing** [comment]
+1 for adding this functionality.
We are using Environments and then we have these one-off `TEST_ENV1_ROOT_URL` variables that are awkwardly in our repo and it makes composable actions harder.

**Comment by stewartadam** [comment]
This is made particularly awkward for using environments with re-usable workflows because `environment` property on jobs is not supported if `using` is present. So as best I can tell isn't a way to load environment secrets in a triggered workflow and pass them to callable workflows, that aren't themselves aware of the environments.

**Comment by DavidPriceNBS** [comment]
+1

**Comment by jzila** [comment]
I have a similar use case - I use `jobs` with a `matrix` to deploy all my containers using the same job spec, and they all need environment secrets. Unfortunately because of the job-environment-deployment coupling, this creates <n> deployments for every workflow run, and they're effectively never "active".

A workflow-level `environment` spec would allow all the nice concurrency/DRY semantics of jobs to apply, while only creating one deployment and allowing all jobs to access environment-specific variables and secrets.

**Comment by gguillen** [comment]
+1 Need this

**Comment by gavinclarkeuk** [comment]
Definitely need a solution to this. We have a multi job deployment process which we use reusable workflows for. Because of the (seemingly arbitrary) limitation on setting `environment` on jobs if `using` is present we can't set the environment in any meaningful way that doesn't result in a single deployments being treated as multiple deployments.

We need to be able to either set `environment` at a workflow level, or lift the restriction on setting it on a job when calling a reusable workflow.

For our use case we aren't really using environment secrets, because the lack of org wide environments makes that impractical. However we do want to use environment protection rules and approvals. The extra UI feedback would be useful too.

**Comment by markfickett** [comment]
This also affects timers. We deploy to our production environment after a 24h timer. (Each environment also has some of its own secrets and variables.)

However, the prod deployment consists of `deploy backend -> migrate backend db` and `build frontend w/ production config -> deploy frontend`, four jobs with two `needs`. So the 24h timer waits once for the `deploy backend` and `build frontend` steps, and then again for the `migrate db` and `deploy frontend` steps. This is similar when manually starting the waiting production jobs if we want to deploy right away; I have to click 'start waiting jobs', wait for the first two `production` jobs to complete, and then click 'start waiting jobs' again.

Instead, I want the timer to run once for everything the workflow is going to do in the production environment. So I would merge to `main`, the `staging` environment deployment jobs would run, and then 24h later the `production` environment deployment jobs would all run in order without additional waits.

**Comment by gerrycampion** [comment]
I also have this similar problem, but I have a workaround that:
- allows for multiple parallel jobs using the env vars
- requires only a single approval

Using this: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/passing-information-between-jobs

You can replace this code:
```yaml
jobs:
  job1:
    runs-on: ubuntu-latest
    environment: PROD
    steps:
      run: echo ${{ vars.VAR1 }}
  job2:
    runs-on: ubuntu-latest
    environment: PROD
    steps:
      run: echo ${{ vars.VAR2 }}
```
With this code:
```yaml
jobs:
  copy_vars:
    name: Copy environment vars from Github Env Vars to outputs to avoid multiple approvals
    runs-on: ubuntu-latest
    environment: PROD
    outputs:
      VAR1: ${{ steps.vars.outputs.VAR1 }}
      VAR2: ${{ steps.vars.outputs.VAR2 }}
    steps:
      - id: vars
        run: |
          echo "VAR1=${{ vars.VAR1 }}" >> "$GITHUB_OUTPUT"
          echo "VAR2=${{ vars.VAR2}}" >> "$GITHUB_OUTPUT"
  job1:
    runs-on: ubuntu-latest
    needs: copy_vars
    steps:
      run: echo ${{ needs.copy_vars.outputs.VAR1 }}
  job2:
    runs-on: ubuntu-latest
    needs: copy_vars
    steps:
      run: echo ${{ needs.copy_vars.outputs.VAR2 }}
```

If you do `copy_vars` and pass the vars as inputs to a reusable workflow containing job1 and job2, then you no longer need `needs: copy_vars` on each job.
There may be a way to generalize this for any number of env vars using something like jq? Any tips to generalize or simplify this would be nice.

**Reply by gavinclarkeuk** [reply]
Whilst this works for extracting the environment variables, it still isn't ideal because if one of the deployment jobs fails the environment deployment will still look at though it succeeded because the copy_vars worked.

**Comment by sphayden** [comment]
+1 to this

**Comment by MikeMcTiernan** [comment]
Is there any official response to this request? This is one of the silliest issues with using environments in Github Actions. If I want my deployments to be efficient and I split the different steps involved into concurrently running jobs, it ends up triggering all of the environment deployment triggers over and over and over again. The outcome of this is users like us having to choose between:

1️⃣ Putting our entire deployment process into a single job that runs in serial, drastically increasing our build and deployment time.

2️⃣ Living with a single workflow triggering multiple deployments--which may involve multiple approval gates, multiple deployments getting created in the UI, a deployment appearing as successful before it's actually completed all of its jobs, etc.

3️⃣ Doing some preliminary step to extract all of the environment's variables to use in subsequent jobs that don't have the environment specified.

All of these are terrible options and most of GH's alternatives have already solved this issue--see Gitlab, Azure DevOps, etc. And to boot, this honestly seems like such a standard use case--not an edge--of this feature that I'm shocked it hasn't been addressed yet.

**Comment by JoshuaRowland** [comment]
This represents a significant gap in functionality.

The fact that this discussion has remained unresolved for over three years raises serious concerns about the long-term viability of the broader feature. It’s making us reconsider whether relying on GitHub Deployments is the right choice compared to using a dedicated deployment tool.

**Comment by chrisponton** [comment]
This is so painful. It's unbelievable that GitHub can't handle something so simple these days. There are numerous AI updates etc, but I can't even perform something basic for mature CI/CD. 

This is from GitHub, the last time I asked:

> This is a planned functionality and Engineering are looking at allowing the use of environments without automatically creating deployments. We don't, however, have a definite timeline for when the feature will be published, sorry. We do expect it to be available in Quarter 4 of this year but that is subject to change. We'll announce it on the [Changelog](https://github.blog/changelog/) section of our blog once it is published.

`but that is subject to change`... like saying you are going to do something, but you aren't...

**Comment by Cube707** [comment]
We also would benefit from this. Our workflow consists of three pars: build, test, deploy. Build requires enviroment-secrets to sign and encrypt parts of the build, and deploy need secrets for deploying. 

Currently the enviroment-deployment looks done once the build finishes and doesn't care about the tests and using aproval for the enviroment doesn't work as each test job needs to be approved separately...

**Comment by MikeMcTiernan** [comment]
🎉 https://github.blog/changelog/2026-03-19-github-actions-late-march-2026-updates/#github-actions-now-allows-developers-to-use-environments-without-auto-deployment

> GitHub Actions now supports using environments in workflows without automatically creating a deployment. Many users have indicated that they want to be able to use the secrets and variables management capabilities of environments without creating additional deployments. If you configure an environment with the deployment: false key, you can access the environment but won’t create a new deployment. More information can be found in [our docs on using environments without deployments](https://docs.github.com/actions/how-tos/deploy/configure-and-manage-deployments/control-deployments#using-environments-without-deployments).

> Please note that if you are using a [custom deployment protection rule](https://docs.github.com/actions/how-tos/deploy/configure-and-manage-deployments/create-custom-protection-rules), you will not be able to use the deployment: false key and using environments will require an auto-deploy.

**Comment by CodeGat** [comment]
That's nice, although you do still have to approve multiple times for each job that uses the environment.
